nixos/modules/services/networking/prosody.nix

```diff
@@ -1,885 +1,659 @@
-            http_upload_file_size_limit = ${cfg.uploadHttp.uploadFileSizeLimit}
-            http_upload_expire_after = ${cfg.uploadHttp.uploadExpireAfter}
-            ${lib.optionalString (cfg.uploadHttp.userQuota != null) "http_upload_quota = ${toLua cfg.uploadHttp.userQuota}"}
-            http_upload_path = ${toLua cfg.uploadHttp.httpUploadPath}
-      ''}
-
-      ${ lib.concatStringsSep "\n" (lib.mapAttrsToList (n: v: ''
-        VirtualHost "${v.domain}"
-          enabled = ${boolToString v.enabled};
-          ${ optionalString (v.ssl != null) (createSSLOptsStr v.ssl) }
-          ${ v.extraConfig }
-        '') cfg.virtualHosts) }
-    '';
+            If you know what you're doing, you can disable this warning by
+            setting config.services.prosody.xmppComplianceSuite to false.
+          '';
+        };
+      in
+      [
+        (mkAssertion "muc" ''
+          You need to setup at least a MUC domain to comply with
+          XEP-0423
+        '')
+
+        (mkAssertion "http_upload" ''
+          You need to setup the http_upload module through
+          config.services.prosody.components to comply with
+          XEP-0423.
+        '')
+      ];
 
-    users.users.prosody = mkIf (cfg.user == "prosody") {
-      uid = config.ids.uids.prosody;
-      description = "Prosody user";
-      inherit (cfg) group;
-      home = cfg.dataDir;
+    warnings =
+      optionals (cfg.modules != { }) [ "The option `services.prosody.modules' has been and split into two separate options: `services.prosody.settings.modules_enabled' and `services.prosody.settings.modules_enabled'." ] ++
+      mapAttrsToList (k: v: "The option `services.prosody.virtualHosts.${k}.enabled' has been renamed to `services.prosody.virtualHosts.${k}.settings.enabled'.") (filterAttrs (_: v: v.enabled != null) cfg.virtualHosts) ++
+      mapAttrsToList (k: v: "The option `services.prosody.virtualHosts.${k}.ssl' has been renamed to `services.prosody.virtualHosts.${k}.settings.ssl'.") (filterAttrs (_: v: v.ssl != null) cfg.virtualHosts)
+    ;
+
+    services.prosody.settings = {
+      log = mkDefault { warn = "*syslog"; error = "*syslog"; };
+      data_path = mkForce cfg.dataDir;
+      network_backend = "event";
+      pidfile = "/run/prosody/prosody.pid";
+
+      authentication = mkDefault "internal_hashed";
+      reload_modules = mkIf (acmeHosts != []) [ "tls" ];
+
+      modules_enabled = [
+        # required for compliance with https://compliance.conversations.im/about/
+        "dialback"
+        "disco"
+        "roster"
+        "saslauth"
+        "tls"
+
+        # not essential, but recommended
+        "blocklist"
+        "bookmarks"
+        "carbons"
+        "cloud_notify"
+        "csi"
+        "pep"
+        "private"
+        "vcard_legacy"
+
+        # nice to have
+        "mam"
+        "ping"
+        "register"
+        "smacks"
+        "time"
+        "uptime"
+        "version"
```

Comment:
Those categories are rather arbitrary. Are they based on something?

Reply:
if i had to guess i would say at one point there were based on something like [this](https://github.com/bjc/prosody/blob/master/prosody.cfg.lua.dist)...? but all i did was preserve the existing behaviour of the module...